### PR TITLE
Enable scalar tensors inside models/DAGs/MediaPipe graphs

### PR DESCRIPTION
**Pull request by dkalinowski**
- enable serving models with inputs/outputs with dimension count=0 (scalars)
- enable serving DAGs with such models
- allow returning scalars from custom nodes
- remove restrictions from DAG validation for scalar connections
- add unit tests

NOTE: tensors with 0 dimensions have default layout "...", whereas regular tensors still keep default layout "N..."

JIRA: CVS-113074

### CODE REVIEW

File: src/test/ensemble_flow_custom_node_tests.cpp [hunk 1]
```diff
@@ -3278,6 +3447,30 @@ TEST_F(EnsembleConfigurationValidationWithDemultiplexer, DemultiplexerNodeNotEno
     ASSERT_EQ(pipelineDefinition->validate(manager), StatusCode::PIPELINE_NOT_ENOUGH_SHAPE_DIMENSIONS_TO_DEMULTIPLY);
 }
 
+TEST_F(EnsembleConfigurationValidationWithDemultiplexer, DemultiplexerNodeNotEnoughDimensionsToDemultiply_Scalar) {
```
Comment (atobiszei) [inline src/test/ensemble_flow_custom_node_tests.cpp:3450]: What is the behavior when someone has DAG with scala input/output, and then enables demultiplexer? What are the metadata and does infer work?

**Review by dtrawins — APPROVED** [review]

File: src/capi_frontend/inferencetensor.cpp [hunk 2]
```diff
@@ -28,8 +28,11 @@ InferenceTensor::InferenceTensor(InferenceTensor&& rhs) :
     shape(std::move(rhs.shape)),
     buffer(std::move(rhs.buffer)) {}
 InferenceTensor::InferenceTensor(OVMS_DataType datatype, const int64_t* shape, size_t dimCount) :
-    datatype(datatype),
-    shape(shape, shape + dimCount) {}
+    datatype(datatype) {
+    if (dimCount > 0) {
+        this->shape = signed_shape_t(shape, shape + dimCount);
```
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/capi_frontend/inferencetensor.cpp:33]: TODO: check what happens for vector constructor with nullptr

File: src/dags/node_library_utils.cpp [hunk 3]
```diff
@@ -115,13 +115,6 @@ Status createTensorInfoMap(struct CustomNodeTensorInfo* info, int infoCount, std
     // At this point it is important to not exit before we iterate over every info object.
     // This is due to a fact that we need to ensure to free resources allocated by shared library using freeCallback.
     for (int i = 0; i < infoCount; i++) {
-        if (info[i].dims == nullptr) {
-            continue;
-        }
-        if (info[i].dimsCount == 0) {
-            freeCallback(info[i].dims, customNodeLibraryInternalManager);
```
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/dags/node_library_utils.cpp:122]: Confirmed during live review - not leaking, free in line 130

File: src/test/scalar/1/saved_model.xml [hunk 4]
```diff
@@ -0,0 +1,47 @@
+<?xml version="1.0"?>
+<net name="TensorFlow_Frontend_IR" version="11">
+	<layers>
+		<layer id="0" name="in" type="Parameter" version="opset1">
```
Comment (dkalinowski) [inline src/test/scalar/1/saved_model.xml:4]: change in to something more descriptive

File: src/layout.cpp [hunk 5]
```diff
@@ -59,10 +59,7 @@ std::optional<size_t> Layout::retrieveBatchIndex() const {
     auto batchPos = this->find(BATCH_DIMENSION_LETTER);
     auto etcPos = this->find(ETC_LAYOUT_DELIMETER);
     if (static_cast<std::string>(*this) == UNSPECIFIED_LAYOUT) {
-        // we want to treat ANY layout as having BS on 0 position
-        // otherwise in any case we extract this we have to check
-        // against layout ...
-        return 0;
+        return std::nullopt;
```
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/layout.cpp:62]: We allow `...` layout and simply skip validating batch size in this scenario. Entire shape is validated instead.

File: src/capi_frontend/inferencetensor.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/capi_frontend/inferencetensor.cpp:33]: Reverted, removed if, since range constructor does not copy when first and last are equal

**Review by atobiszei — APPROVED** [review]

**Pull request merged**